### ISSUE TITLE
Cannot always write to net.Socket created in ipcMain handler

### ISSUE THREAD

**Issue by thomasschaub** [electron/electron#8279]
* Electron version: v1.4.13
* Operating system: Ubuntu 16.04

### Expected behavior

After successfully creating a socket in an `ipcMain` handler, everything written to the socket should be transferred (more or less) right away.

### Actual behavior

After successfully creating a socket in an `ipcMain` handler, everything written to the socket in only transferred when that handler is executed for a second time.

### How to reproduce

I wrote a small demo to showcase the behavior:

* Clone https://github.com/thomasschaub/electron-net-test
* `npm install`
* `nc -l 1234 & npm run demo-bad`

In a nutshell, this will work as expected:

    let socket = net.connect(...);
    ipcMain.on('x', () => {
        socket.write(...);
    });

While this won't:

    let socket;
    ipcMain.on('x', () => {
        if (socket === undefined) {
             socket = net.connect(...);
        }
        socket.write(...); // does not write if the socket was just created.
    });

**Comment by MarshallOfSound** [comment]
Without looking into it much it would appear that `net.connect` is asynchronous.  You probably have to wait for the connection to be ready before writing to it.

**Comment by thomasschaub** [comment]
Good point. I added a bit of logging to the `connect` event and apparently, the event is also only triggered when the handler is executed for the second time.

**Comment by kevinsawicki** [comment]
@thomasschaub here is the output I'm seeing with your sample repo:

```
$ npm run demo-bad

> electron-net-test@1.0.0 demo-bad /Users/kevin/github/electron-net-test
> NODE_DEBUG=net DEBUG=diy-socket electron . 3

socketCreationPoint == 3
NET 22254: createConnection [ { port: 1234 }, [Function] ]
NET 22254: pipe false undefined
NET 22254: connect: find host localhost
NET 22254: connect: dns options { family: undefined, hints: 1024 }
  diy-socket start +0ms
  diy-socket end +1ms
NET 22254: _read
NET 22254: _read wait for connection
NET 22254: afterConnect
  diy-socket Socket connected +3ms
  diy-socket write completed +0ms
NET 22254: _read
NET 22254: Socket._read readStart
```

Looks like `write completed` is logged, what output are you seeing?

**Comment by thomasschaub** [comment]
This shows a session where I click twice:

At app start (before any clicks):

     % nc -l 1234 & npm run demo-bad
    [1] 5596

    > electron-net-test@1.0.0 demo-bad /home/thomas/Dev/electron-net-test
    > NODE_DEBUG=net DEBUG=diy-socket electron . 3

    socketCreationPoint == 3

After first click:

    NET 5614: createConnection [ { port: 1234 }, [Function] ]
    NET 5614: pipe false undefined
    NET 5614: connect: find host localhost
    NET 5614: connect: dns options { family: undefined, hints: 32 }
      diy-socket start +0ms
      diy-socket end +2ms
    NET 5614: _read
    NET 5614: _read wait for connection

After second clicks:

      diy-socket start +12s
      diy-socket end +1ms
    NET 5614: afterConnect
      diy-socket Socket connected +1ms
    hey
    hey
      diy-socket write completed +2ms
    NET 5614: _read
    NET 5614: Socket._read readStart
      diy-socket write completed +1ms

As you can see, `write completed` is logged, but only after the second click.

Which OS are you using? A bit of context: I got an Electron app that uses some JSON RPC library. The requests are initiated in a ipcMain handler (similar what happens in my example). This worked fine on Windows, but didn't on Ubuntu 16.04. The JSON library uses Node's http module, which in turn uses Node's socket module. I stopped digging there and wrote the sample app (and rolled my own JSON RPC using Electron's net module, which works as expected, but it's something I'd rather not have to do). Anyhow, this is why I think it might not be reproducible on OS X or maybe even other Linux distros.

**Comment by soomtong** [comment]
I had same issue like this.
only linux makes this result, they didn't callback when socket connected or didn't connect at first time.
but Windows and MacOS worked we expected.

And I tried this code in Render Process not Main Process without IPC call. 
It just works. then I moved my code to renderer process.

**Comment by zcbenz** [comment]
I'm merging this to #1833 because they are essentially the same bug with our Node integration implementation.
